### PR TITLE
[JUJU-1210] Rewrite to bash-based nw-deploy-xenial-arm64-lxd

### PR DESCRIPTION
**Pull request by anvial**
This patch provides changes in the deploy suite (bash-based testing system) to be able to deploy charms on arm64 machines at the same time without breaking tests when this suite is executed on amd64 machines.

IMPORTANT: I've disabled following tests: run_deploy_lxd_to_container, run_deploy_lxd_profile_charm_container, run_deploy_local_lxd_profile_charm, because we have known issue that breaks lxd-in-containers (on lxd). So, these tests commented only on passing QA steps. I'll uncomment them before merging. 

We plan this test to be targeted on AWS (on "ephemeral-focal-4c-16g-arm64" mode) in our Jenkins.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [ ] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

```sh
# bootstrap aws cloud with any juju you have
juju bootstrap aws aws
juju add-machine --constraints "arch=arm64 cores=2 root-disk=32G"
# wait for the machine to be started
juju ssh 0 
# inside the machine
sudo apt install make cmake python3-pip libssl-dev jq shellcheck
# install juju, based on this PR
git clone https://github.com/juju/juju.git
cd juju
git fetch origin pull/14409/head:test-branch
git checkout test-branch
make install-dependencies
make build
export GOPATH="/home/ubuntu/go:/home/ubuntu/juju/_build/linux_arm64"
export PATH=$PATH:$GOROOT/bin:$GOPATH/bin
# init localhost cloud
sudo newgrp lxd
exit
sudo adduser $USER lxd
newgrp lxd
lxd init --auto
lxc network set lxdbr0 ipv6.address none
# launch deploy_charms tests
cd tests
export BUILD_ARCH=arm64
./main.sh -v -p lxd deploy test_deploy_charms
```

### CODE REVIEW

File: tests/includes/juju.sh [hunk 1]
```diff
@@ -289,6 +317,9 @@ post_add_model() {
 		add_images_for_vsphere
 		;;
 	esac
+
+	arch=$(uname -m)
```
Comment (wallyworld) [inline tests/includes/juju.sh:321]: we shouldn't be setting arch based on the machine running the test
arch needs to be explicitly passed in as a test parameter; if we say we're testing on arm64, then we pass that as a test arg and explicitly use it when deploying etc

File: tests/suites/deploy/deploy_charms.sh [hunk 2]
```diff
@@ -213,7 +213,7 @@ run_deploy_lxd_to_container() {
 		attempt=$((attempt + 1))
 		if [ $attempt -eq 10 ]; then
 			# shellcheck disable=SC2046
-			echo $(red "timeout: waiting for lxc profile to show 50sec")
+			echo $(red "timeout: waiting for removal of lxc profile 50sec")
```
Comment (wallyworld) [inline tests/suites/deploy/deploy_charms.sh:216]: why this change?

Comment (anvial) [reply tests/suites/deploy/deploy_charms.sh:216]: Ian, I suppose tests run_deploy_lxd_to_machine and run_deploy_lxd_to_container check the same issue. If so, in the second "while-cycle" we need to ensure that the old profile is removed, and I've just fixed the error message to to reflect it.

**Comment by hmlanigan** [comment]
@anvial I have nothing to add beyond what @wallyworld mentioned.

File: tests/includes/juju.sh [hunk 3]
```diff
@@ -289,6 +289,8 @@ post_add_model() {
 		add_images_for_vsphere
 		;;
 	esac
+
+	juju set-model-constraints "arch=${BUILD_ARCH:-}"
```
Comment (wallyworld) [inline tests/includes/juju.sh:293]: It's not really "build arch" though is it? This env var is set in the Makefile when building juju but this is totally unrelated to running tests.
And what if it is not set? Constraints parsing will fail.

**Review by wallyworld — COMMENTED** [review]
Using an env var is the right approach. But not BUILD_ARCH, as that is not relevant here at all.
We could use WORKLOAD_ARCH.
We really need to look at introducing a new suite, "multiarch" or something like that. This suite will have tests which look at BOOSTRAP_PROVIDER and WORKLOAD_ARCH. If the specified arch is supported by the cloud, the run the test, else skip. eg on AWS or GCE or Azure we can test amd64 and arm64 but not s390x etc. We'd test s390x with our MAAS once that is set up etc.
The suite can initially have one test, eg "test_charm" where a charm for the specified arch is deployed. The WORKLOAD_ARCH will be used to set deployment constraints (not model). We can then look to extend things by having another test where model constraints are set, eg using MODEL_ARCH and see that this works. And we can have both set to ensure that on AWS for example, we can deploy both amd64 and arm64 charms with one being set by the model constraint and the other by the deployment constraint.
Using a new suite allows tests specifically designed for multi-arch testing to be written.

File: tests/includes/juju.sh (hunk 3 above)
Comment (anvial) [reply tests/includes/juju.sh]: Taking into consideration your comment below, I will:

1. Check MODEL_ARCH and use it as a model constraint if defined
2. Check WORKLOAD_ARCH and use it as a deployment constraint if defined

File: tests/suites/deploy/deploy_charms.sh [hunk 4]
```diff
@@ -235,14 +235,14 @@ test_deploy_charms() {
 
 		run "run_deploy_charm"
 		run "run_deploy_specific_series"
-		run "run_deploy_lxd_to_container"
-		run "run_deploy_lxd_profile_charm_container"
+		#		run "run_deploy_lxd_to_container"
+		#		run "run_deploy_lxd_profile_charm_container"
```
Comment (wallyworld) [inline tests/suites/deploy/deploy_charms.sh:239]: fix me

**Review by wallyworld — COMMENTED** [review]
Needs checks that the MODEL_ARCH is compatible with the controller cloud

**Review by wallyworld — APPROVED** [review]
We can land to get the test green and iterate
(after comments are removed)
